src/Umbraco.Core/Composing/Composers/FileSystemsComposer.cs

```diff
@@ -76,15 +76,13 @@ public static IContainer ComposeFileSystems(this IContainer container)
             // register IFileSystems, which gives access too all filesystems
             container.RegisterSingleton<IFileSystems>(factory => factory.GetInstance<FileSystems>());
 
-            // register IMediaFileSystem
-            var virtualRoot = GetMediaFileSystemVirtualRoot();
-            container.RegisterSingleton<IFileSystem>("media", factory => new PhysicalFileSystem(virtualRoot));
-            container.RegisterSingleton<IMediaFileSystem>(factory => factory.GetInstance<FileSystems>().GetFileSystem<MediaFileSystem>());
+            // register IMediaFileSystem with its actual, underlying filesystem factory
+            container.RegisterSingleton<IMediaFileSystem>(factory => factory.GetInstance<FileSystems>().GetFileSystem<MediaFileSystem>(MediaInnerFileSystemFactory));
```

Comment:
This looks clunky. I would expect a cleaner endpoint that allows you to register the two type simultaneously.

Reply:
It does bother me though that I have two types to register. The whole `FileSystemWrapper` API also seems superfluous only calling the inner `IFileSystem` implementation.

Reply:
I would expect to be able to entirely replace the `IMediaFileSystem` used (by default) in Umbraco. The `FileSystems` component should expose _my_ `IMediaFileSystem`. It should be **late bound** so it can be configured (differently) in json or xml. Since configuration with late binding in MS.DI is likely to be omitted and/or difficult, could this be:

    container.RegisterSingleton<IMediaFileSystem>(InstantiateConfiguredFileSystemWrappedInShadowWtfUsingReflection)

?

Reply:
I don't think it should be configurable via config files at all. If you create a diaspora of configuration points you open yourself up to temporal anomalies.  

Plus of course you end up having to search for the correct location as to where things should be configured.

Reply:
I think there's a broader goal of allowing people who for some reason don't want to learn about DI to use "the old ways". AFAIK, the file system is the only thing that's late bound today, so I wouldn't mind keeping it that way. That's not to say that you shouldn't be able to override the default in whichever way you want.

Reply:
I'd like to wire this plug but I don't want to learn how to wire this plug.

Reply:
There are a lot of them out there. 😱

Reply:
It _is_ clunky. Doing incremental updates here, starting from a ... interesting ... codebase.

Pain points:

(a) `IMediaFileSystem` implements more than just `IFileSystem` and so in v7 it's something that wraps a real `IFileSystem` - so what you could configure is the real `IFileSystem` and it would be wrapped. I'd love to do it differently.

(b) _all_ filesystems end up being wrapped somehow, because we implement "shadowing" - and what's shadowing? it's some sort of filesystem-level transaction that's required by deploy. when "shadow" is on, every filesystem operation actually happens in a temp dir, and is committed (or not) at the end.

Not pretending what we have now is all clean. Accepting changes ;-)

Reply:
Is shadowing the byproduct of current API constraints?

Reply:
Shadowing means that at some point, Core does not talk to your `IFileSystem` directly but to some wrapper. That is, transactions are implemented *on top* of `IFileSystem`. So they can work whether on a physical fs, or Azure or S3 or...

So you need to provide a filesystem for medias, eg a physical one, or a remote one, or... either via config (in v7) or code (in v8 - haven't entirely removed the config option but would love to). And we wrap it and implement shadowing on top of it.

This is what makes it all weird. Cannot do

```container.Register<IMediaFileSystem>(_ => new PhysicalFileSystem(path));```

and inject `IMediaFileSystem` where needed, because we want to wrap it (or decorate, as @lars-erik  would say). Hence the currently weirdish syntax where `IMediaFileSystem` is obtained through `FileSystems`.

None of this is ideal - looking for ideas.

Reply:
And for @lars-erik : no, cannot *entirely* swap `IMediaFileSystem` with yours because ... yours would have to implement shadowing etc.

Reply:
So this is exactly what I've been talking about when I talk about interception. Here's a suggestion:

Add another IUmbracoComponent artifact that depends on IUmbracoUserComponent. In it you use interception to decorate all registered IFileSystem components with ShadowFileSystem. This means that packages and consumers can override the media file system, add their own and generally do as they please. When everything is ready, you decorate them all with whatever you want. Shadowing, auditing, logging, whatnot.

Reply:
You can even make the decoration a plugin that only registers when deploy is present.

Reply:
Which I just realize will be impossible without generics constraints or named services. What could be done is make another interface for implementors of IFileSystem, ie. IPluginFileSystem. Everything is registered as that type, and then decorated with IFileSystem(s) implementing whatever Umbraco needs. Everyone has to know they have to inject IFileSystems and leave IPluginFileSystems "alone". A bit more messy, but the only way (?) around it without any other interception mechanism.

Reply:
Nah.... Too complex.  You have a Shadow class (probably should be renamed to something more accurate) and pass the `IFileSystem` or `IMediaFileSystem` to it via the constructor. Your Shadow class does anything custom that requires transactions/temp storage and implementations are free to come and go.

Reply:
Many facets to this. Just to supply the quick solution side of the discussion: As long as the FS components have singleton lifetimes, they can be decorated. Whatever's registered by the last user composition can be resolved, decorated, and then re-registered. Will work, but only for singletons.

Reply:
Here's the real solution. :)

@JimBobSquarePants googled this for me a while ago wrt. interception. I stumbled across it again now to quickly solve autodiscovery. I suggest we bring it in as salvation.
It can also do .Decorate<OpenGeneric<>>(), which is what you want for the deploy thingy.

https://github.com/khellang/Scrutor